src/settings.rs

```diff
@@ -27,22 +34,156 @@ impl<'src> Settings<'src> {
         .collect()
+    } else if let Some(shell_args) = &config.shell_args {
+      shell_args.iter().map(String::as_ref).collect()
+    } else if cfg!(windows) && self.windows_powershell {
+      WINDOWS_DEFAULT_SHELL_ARGS.to_vec()
+    } else {
+      DEFAULT_SHELL_ARG.to_vec()
+    }
+  }
+}
+
+#[cfg(test)]
+mod tests {
+  use crate::setting::Shell;
+
+  use super::*;
+
+  #[test]
+  fn default_shell() {
+    let mut settings = Settings::new();
+    settings.windows_powershell = false;
+
+    let config = Config {
+      shell_command: false,
+      ..testing::config(&[])
+    };
+
+    assert_eq!(settings.shell_binary(&config), "sh");
+    assert_eq!(settings.shell_arguments(&config), vec!["-cu"]);
+  }
+
+  #[test]
+  fn default_shell_powershell() {
+    let mut settings = Settings::new();
+    settings.windows_powershell = true;
+
+    let config = Config {
+      shell_command: false,
+      ..testing::config(&[])
+    };
+
+    if cfg!(windows) {
+      assert_eq!(settings.shell_binary(&config), "powershell.exe");
+      assert_eq!(
+        settings.shell_arguments(&config),
+        vec!["-NoLogo", "-Command"]
+      );
+    } else {
+      assert_eq!(settings.shell_binary(&config), "sh");
+      assert_eq!(settings.shell_arguments(&config), vec!["-cu"]);
+    }
+  }
+
+  #[test]
+  fn overwrite_shell() {
+    let mut settings = Settings::new();
+    settings.windows_powershell = false;
+
+    let config = Config {
+      shell_command: true,
+      shell: Some("lol".to_string()),
+      shell_args: Some(vec!["-nice".to_string()]),
+      ..testing::config(&[])
+    };
+
+    assert_eq!(settings.shell_binary(&config), "lol");
+    assert_eq!(settings.shell_arguments(&config), vec!["-nice"]);
+  }
+
+  #[test]
+  fn overwrite_shell_powershell() {
+    let mut settings = Settings::new();
+    settings.windows_powershell = true;
+
+    let config = Config {
+      shell_command: true,
+      shell: Some("lol".to_string()),
+      shell_args: Some(vec!["-nice".to_string()]),
+      ..testing::config(&[])
+    };
+
+    assert_eq!(settings.shell_binary(&config), "lol");
+    assert_eq!(settings.shell_arguments(&config), vec!["-nice"]);
+  }
+
+  #[test]
+  fn shell_cooked() {
+    let mut settings = Settings::new();
+    settings.windows_powershell = false;
+
+    settings.shell = Some(Shell {
+      command: StringLiteral {
+        kind: StringKind::from_token_start("\"").unwrap(),
+        raw: "asdf.exe",
+        cooked: "asdf.exe".to_string(),
+      },
+      arguments: vec![StringLiteral {
+        kind: StringKind::from_token_start("\"").unwrap(),
+        raw: "-nope",
+        cooked: "-nope".to_string(),
+      }],
+    });
+
+    let config = Config {
+      shell_command: false,
+      ..testing::config(&[])
+    };
+
+    assert_eq!(settings.shell_binary(&config), "asdf.exe");
+    assert_eq!(settings.shell_arguments(&config), vec!["-nope"]);
+  }
+
+  #[test]
+  fn shell_args_present_but_not_shell() {
```

Comment:
For completeness, it would be good to see something like `shell_present_but_not_shell_args`